Fixes LoggerSilence#silence threadsafety

rabbitt opened the pull request:
- uses Logger instance defined level if no custom local log level defined
- Keeps track of local log level per logger instance + thread id
- prevents memory leakage by removing local level hash key/value on #silence method exit
- Test case for threadsafety issue added (and passes with these changes), based @rdubya's tests
- fixes #20490 and supersedes https://github.com/rails/rails/pull/16885

rabbitt committed:
Fixes LoggerSilence#silence threadsafety
- uses instance defined level if no custom local log level defined
- Keeps track of local log level per [ thread + object-instance ]
- prevents memory leakage by removing local level hash key/value on #silence method exit
- avoids the use of Thread local variables

rabbitt commented:
@rafaelfranca here's the PR. I would have sent this yesterday but ended up swamped with work stuff.

kaspth commented on activesupport/lib/active_support/logger.rb:
```diff
@@ -46,6 +46,31 @@ def initialize(*args)
+      Thread.current[:logger_level] || Thread.main[:logger_level] || super
+    end
+
+    def level=(l)
```
Let's write out the variable name.

kaspth commented on activesupport/lib/active_support/logger.rb:
```diff
@@ -46,6 +46,31 @@ def initialize(*args)
+    end
+
+    def add(severity, message = nil, progname = nil, &block)
+      return true if @logdev.nil? or (severity || UNKNOWN) < level
```
Please stick to `||`.

kaspth commented on activesupport/test/logger_test.rb:
```diff
@@ -16,6 +16,21 @@ def setup
   end
 
+  def with_level(level = Logger::INFO)
+    begin
```
Methods are enclosed in their own begin scope, so you don't need to call begin and ensure can be at the same indentation as def.

kaspth commented on activesupport/test/logger_test.rb:
```diff
@@ -16,6 +16,21 @@ def setup
     @logger  = Logger.new(@output)
   end
 
+  def with_level(level = Logger::INFO)
```
It doesn't really make sense with a default argument here. What does `with_level(nil)` mean? Let's go explicit.

kaspth commented on activesupport/test/logger_test.rb:
```diff
@@ -120,14 +135,57 @@ def test_buffer_multibyte
+  def test_logger_thread_safety
+    @logger.level = Logger::INFO
+
+    assert @logger.level == Logger::INFO,
```
Why aren't we using `assert_equal` throughout this test?

kaspth replied:
Ah, I see. Because the `level` returns an integer and not a symbol. I'd rather we added a custom and private assertion, like `assert_level`.

rabbitt commented:
@kaspth ping.

kaspth commented on activesupport/test/logger_test.rb:
```diff
@@ -120,14 +120,73 @@ def test_buffer_multibyte
+  end
+
+  def assert_level(level)
+    assert @logger.level == level, "Expected level #{level_name(level)}, got #{level_name(@logger.level)}"
```
I think we can switch to `assert_equal` now.

rabbitt replied:
why don't we meet halfway and I'll add assert_equal but keep the more explicit (and readable) assert message above instead of relying on the diff output that assert_equal generates.

kaspth commented on activesupport/test/logger_test.rb:
```diff
@@ -120,14 +120,73 @@ def test_buffer_multibyte
+  def assert_level(level)
+    assert @logger.level == level, "Expected level #{level_name(level)}, got #{level_name(@logger.level)}"
+  end
+
```
:scissors: this line.

kaspth commented on activesupport/test/logger_test.rb:
```diff
@@ -120,14 +120,73 @@ def test_buffer_multibyte
+  end
+
+  private
+
```
:scissors: this line and indent under private.

kaspth commented on activesupport/test/logger_test.rb:
```diff
@@ -120,14 +120,73 @@ def test_buffer_multibyte
     assert @output.string.include?("THIS IS HERE")
   end
+
+  def test_logger_thread_safety
```
Actually we're testing that the logger level is thread safe :wink:

kaspth commented on activesupport/test/logger_test.rb:
```diff
@@ -120,14 +120,73 @@ def test_buffer_multibyte
+      #    ~0.3 | 2nd sleep end   | <sleeping>
+      #    ~0.4 | <dead>          | 2nd sleep end
+
+      Thread.new do
```
I still find this test with the threads hard to understand, and I'm scared we might run into timing issues (might be overblown).

Could `stop` be of help here? That's pretty much what you're doing with `sleep`.

matthewd replied:
I haven't looked at the rest of this PR, but you want [latch](https://github.com/rails/rails/blob/34b2eec02c23ec50caae06cd1c95d5d9b3070829/activesupport/lib/active_support/concurrency/latch.rb).

rabbitt commented:
@kaspth: pong.

kaspth commented on activesupport/test/logger_test.rb:
```diff
@@ -120,14 +121,64 @@ def test_buffer_multibyte
+    def assert_level(level)
+      assert_equal level, @logger.level, "Expected level #{level_name(level)}, got #{level_name(@logger.level)}"
+    end
+end
```
New line here :smile:

kaspth commented:
@matthewd does the test using latch look good to you? If so, I'm :+1: on this.

rabbitt commented:
@matthewd what's the good word? :-)

rabbitt commented:
@matthewd, @kaspth any chance this could get looked at today? We (at my company) are holding off on monkey patching our own code to evade the threadsafety issue until we know for sure what the final patch upstream (here) looks like, to reduce future upgrade issues. 

Optionally, if someone could just confirm that the main substance (sans test) of this patch will make it in, that would free up our developers to backport this patch into our code in an upgrade-safe way.

Thanks!

rabbitt commented:
Any word on this? It's been over a week now......

rabbitt commented:
@matthewd, @kaspth, @rafaelfranca - It's been exactly three weeks since I've heard from anyone. What's the word on this? I understand that you guys are doing this in your spare time (so am I!), so I know that it can be hard to keep up at times. However, three weeks is a bit excessive in waiting for a response. A simple "we're a bit busy right now but haven't forgotten" would be nice and go a long way towards making me feel like I'm not wasting my time (I didn't *have* submit a patch after all)...

rabbitt commented:
*sigh* ping...

shakycode commented:
@rabbitt I'm sure they will get to it.  Did the test you write break CI?  Maybe you can get it to pass in the meantime?  Patience, Daniel-San. :+1:

rafaelfranca commented:
@tenderlove @matthewd @thedarkone What do you guys think?

rafaelfranca commented:
Yeah, the test failure seems to be related.

thedarkone commented on activesupport/lib/active_support/logger.rb:
```diff
@@ -46,6 +46,31 @@ def initialize(*args)
+    end
+
+    def add(severity, message = nil, progname = nil, &block)
+      return true if @logdev.nil? || (severity || UNKNOWN) < level
```
Why are you checking `@logdev.nil?`? Isn't it an i-var Rails has no control over?

rabbitt replied:
ActiveSupport::Logger inherits from Logger, so it does have "control" over ``@logdev``. The reason it's checked here is because it's replicating the behavior of the super method. I personally see no problem short-circuiting here instead of adding another method call to the stack (via ``super``) but, if you feel super strongly about it, I'll be honest - I don't care, and will remove it.

thedarkone commented on activesupport/test/logger_test.rb:
```diff
@@ -120,14 +121,64 @@ def test_buffer_multibyte
+
+        # allow thread 1 to finish
+        thread_1_latch.release if thread_number == 2
+        assert_level(Logger::INFO)
```
These two lines should be swapped, this way the check for the `info` logger level is done while `thread 1` is still guaranteed to be in `with_level(:error)`.

thedarkone commented:
@rabbitt I'm not familiar with how web-servers/background job processors treat `Thread.main`. Is it guaranteed not to perform any work? Because otherwise its logger level changes are system-wide and will affect other threads.

Also, there is slight inconsistency with how `Thread.main` acts as a default authoritative source of logger level, but once a non-main thread uses `LoggerSilence#silence` it gains its own "sticky" thread level (before `silence` its `Thread.current[:logger_level] == nil`, while afterwards `Thread.current[:logger_level] != nil`). However, there is no documented obligation by Rails that all threads must follow `Thread.main`'s logger level, so this might be fine.

rabbitt commented:
@thedarkone I'm not sure if your asking me a question, making a statement or just waxing poetic regarding the use of Thread.main for logger_level. Can you elaborate?

rabbitt commented:
@shakycode, @rafaelfranca CI is passing now.

thedarkone commented:
@rabbitt no, I don't have super strong feelings for your code and no, I'm not trying to make a statement or be poetic about your usage of `Thread.main`. I was asked to review your PR, can you __please__ keep it professional?

Your PR will fail a test like this:
```ruby
def test_logger_level_thread_safety2
  @logger.level = Logger::INFO
  assert_level(Logger::INFO)

  latch  = ActiveSupport::Concurrency::Latch.new
  latch2 = ActiveSupport::Concurrency::Latch.new

  t = Thread.new do
    latch.await
    assert_level(Logger::INFO)
    latch2.release
  end

  with_level(Logger::ERROR) do
    latch.release
    latch2.await
  end

  t.join
end
```

Is the test wrong?

rabbitt commented:
@thedarkone I'm sorry. It's not really excuse - more of a description of my state of mind - but, it's difficult to maintain a high level of decorum after a month and a half of feeling like I've wasted my time on this. I sincerely appreciate that you are looking at this now though and I hope that we can continue to move forward without hurt feelings on your side, or me feeling like I'm wasting my time again.

That being said, I do prefer direct requests and concrete examples of problem code rather than the more recent tip-toeing-around approach that's become the modus vivendi in software development of late. In other words, if you find a problem with my approach - be direct and hit me hard (imagine you're Linus Torvalds). If I firmly believe in my approach, and have data to back it up, I'll do so. If I don't, I'll see myself for the idiot that I am and find another way of tackling the problem.

As for the Thread.main issue - great find. :smile:  The whole purpose of doing that was to attempt to fallback to the applications default level without having to inspect the Rails configuration. Gimme a little while to chew on this and see what I can come up with. Though, if you have any ideas offhand, I'm open to hearing them.

rabbitt commented:
Hi @thedarkone,

(CC: @shakycode, @rafaelfranca, @tenderlove, @matthewd)

I've finally had a moment to take a look at this again and I came up with a solution: instead of relying on the local level set for Thread.main, the code instead sets a default_logger_level on Thread.main the first time ```#level=``` is called. This obviously makes the assumption that the first value set is a good default to use (an assumption that has to be made at some point in order for this to work, imho).

Btw, I additionally realized that my original solution was broken in the case where multiple loggers are instantiated, so to insulate them from each other, I've updated the hash key used to include the ```object_id``` and added an additional test for per object log level thread safety.

Please review and provide feedback.

P.s. All tests (including yours, which I added) are passing.

Thanks!,

--
Carl

thedarkone commented:
Issues with current PR:
 * `default_level_key` and `local_level_key` will leak memory if loggers are created and garbage collected,
 * `object_id` can be reused, so a newly created logger might get the same `object_id` as a garbage collected logger, with the same `object_id` it might inherit the previous logger's custom level,
 * `Thread#[]`/`[]=` methods (ie `Thread.current[local_level_key]`) convert their key args into symbols, this will leak memory if run on MRI without symbol GC.

Unrelated to that, `ActiveSupport::Concurrency::Latch.new` has been deprecated, so test.rb needs to be grep replaced:
 * `require 'active_support/concurrency/latch'` -> `require 'concurrent/atomics'`,
 * `ActiveSupport::Concurrency::Latch.new` -> `Concurrent::CountDownLatch.new`,
 * `latch.await` -> `latch.wait`,
 * `latch.release` -> `latch.count_down`.

-----
Here's what I propose.

Deprecate `LoggerSilence#silence` and provide `LoggerSilence#thread_local_silence` instead. Setting `logger.level` directly changes the log level for all threads, just as before. The only way to temporarily and thread-locally silence a logger is via the new `thread_local_silence` method.

I haven't actually run this, but the idea is that a method taking a block, we can clean up and not leak memory. Other than that, I want to avoid relying on Symbol GC and be as efficient as possible:

```ruby
module LoggerSilence
  extend ActiveSupport::Concern
  
  included do
    cattr_accessor :silencer
    self.silencer = true
  end
  
  def silence(temporary_level = Logger::ERROR)
    ActiveSupport::Deprecation.warn('LoggerSilence#silence is deprecated. Please use LoggerSilence#thread_local_silence instead.')
  end
  
  def level
    if custom_log_levels = Thread.current[:rails_log_levels]
      custom_log_levels[self]
    end || super
  end
  
  # Silences the logger for the duration of the block.
  def thread_local_silence(temporary_level = Logger::ERROR)
    if silencer
      begin
        custom_log_levels = Thread.current[:rails_log_levels] ||= {}

        old_custom_level    = custom_levels[self]
        custom_levels[self] = temporary_level

        yield self
      ensure
        if old_custom_level
          custom_levels[self] = old_custom_level
        elsif custom_levels.size == 1
          Thread.current[:rails_log_levels] = nil
        else
          custom_levels.delete(self)
        end
      end
    else
      yield self
    end
  end
end
```

rabbitt commented:
@thedarkone Though it doesn't actually protect against direct changes to `Logger#level=` itself, I really like your approach as it should, at the very least, handle the  majority of use cases within Rails. 

I've made changes to my PR to match the gist of what you were getting at, though I've removed the Thread Local Variable entirely as it's essentially superfluous with this method, though at the cost of an extra Hash class/instance variable (via: `cattr_accessor :local_levels`). This approach, like yours, should still be generally safe from memory leakage, and fully nestable.

Thoughts?

BrentWheeldon commented:
It looks like the folks over at `activerecord-session_store` have also solved [this](https://github.com/rails/activerecord-session_store/blob/master/lib/active_record/session_store/extension/logger_silencer.rb). We've been using this solution for a while with no issue.

thedarkone commented:
@rabbitt nice idea of having an internal log level map, I haven't thought of that and I really hate adding more thread local keys into threads (`Thread.current[:rails_log_levels]`).

We'll need to replace `@local_levels = {}` with `Concurrent::Map`, so `@local_levels = Concurrent::Map.new(:initial_capacity => 2)` (it is worth it to tell `Concurrent::Map` that it is probably going to a be small map), also don't forget `require 'concurrent'` at the top.

I think `level` can be rewritten like this:
```ruby
def level
  local_levels[local_log_id] || super
end
```

Other than that, this is :+1: from me.

rabbitt commented:
@thedarkone changes made - travis running now. :-)

rabbitt commented:
@thedarkone - one other thing: what do I need to do to get this backported to 4.x (we're using 4.1.x at my company, so it'd be helpful to get it there at least for us) ? Should I just create a PR against both 4-1-stable and 4-2-stable, or ..?

thedarkone commented:
Since backporting is non trivial (see below) yes, you'll need to create additional PRs against other branches. I myself usually wait for the `master` PR to land and only then request backports.

In backports you'll have to go back to `ActiveSupport::Concurrency::Latch.new` (reverting the changes from [my comment](https://github.com/rails/rails/pull/20507#issuecomment-141620393)) and:
 * `require 'concurrent'` -> `require 'thread_safe'`,
 * `Concurrent::Map.new(:initial_capacity => 2)` -> `ThreadSafe::Cache.new(:initial_capacity => 2)`.

PS: @rafaelfranca I think this needs to be backported.

rabbitt commented:
@thedarkone thanks for the insight, and thanks for working with me on this PR. :-)

rabbitt commented:
@rafaelfranca any thoughts on this? :-)

rafaelfranca commented:
R? @rafaelfranca

rabbitt commented:
@rafaelfranca - will this be merged in anytime soon? It's been 3 months now since there was any traction on this (6 months since the initial PR was submitted)...

rafaelfranca commented:
Waiting travis to run and I'll merge

rabbitt commented:
thanks @rafaelfranca

rabbitt commented:
@rafaelfranca looks like it's all green :-)

rafaelfranca commented:
Could you provide the backport PR?

rabbitt commented:
@rafaelfranca Sure, though I'll need a couple of days to work on it.

rabbitt commented:
@rafaelfranca I have some time to work on backporting now - are there any docs on how to proceed with backports? Should I just branch off of the latest tags for 4.0, 4.1, 4.2; how far back should I go? (note: I'd at least like to get 4.1.x in there as that's the version my company is on.)

rafaelfranca commented:
You only need to branch off from the stable branches, and you can go back to 4-1-stable that is the oldest maintained version.

TL;DR:

```
git checkout -b 4-2-stable-fix 4-2-stable
# open PR against 4-2-stable
git checkout -b 4-1-stable-fix 4-1-stable
# open PR against 4-1-stable
```

swapnilabnave commented:
@rabbitt Many thanks for the Logger ThreadSafety fix.
Is this available in rails 3.2.22.1 ?
If not can someone suggest me a way to apply it in my rails application ?

swapnilabnave commented:
Only difference being, I see this issue(logs disappear in production) because of non-thread-safe silencing done in [benchmarkable.rb](https://github.com/rails/rails/blob/3-2-stable/activesupport/lib/active_support/benchmarkable.rb#L48). Which is used while playing with [active_record session-store](https://github.com/rails/rails/blob/3-2-stable/activerecord/lib/active_record/session_store.rb#L306-L345).

swapnilabnave commented:
This patch worked for me. Assuming this wont cause any sort of deadlock or wait timeout.

```ruby
module ActiveSupport
  module Benchmarkable
    extend ActiveSupport::Concern

    mattr_accessor :mutex
    self.mutex = Mutex.new

    def silence
      mutex.synchronize do
        begin
          old_logger_level, logger.level = logger.level, ::Logger::ERROR if logger
          yield
        ensure
          logger.level = old_logger_level if logger
        end
      end
    end
  end
end
```

rabbitt commented:
@swapnilabnave here's a patch for your particular version. It's tested against MRI Ruby 2.1.5 and JRuby 1.7.23: 
   https://gist.github.com/rabbitt/5247188b03f8d92cd7214362022979c6

You'll need to patch your files yourself, if you want to use this. You can patch it by doing something like the following:

cd $(dirname $(gem which active_support/buffered_logger.rb)) 
patch -p1 < <(curl -s https://gist.githubusercontent.com/rabbitt/5247188b03f8d92cd7214362022979c6/raw/9ee617250d52d8eefef6f79d05d5e8167c7abd00/threadsafe-logger-active_support-v3.2.22.1.patch)

Though, ultimately, you might be better off just writing your own logger that's threadsafe to avoid potential issues with monkey-patching, or overwriting original code (note: you could probably get away with importing Rails 4's logger class into your code).

swapnilabnave commented:
Thanks @rabbitt 
I'll write my own logger.
